From 4bb3332b0c92e50149776449f0c502a76f6f2642 Mon Sep 17 00:00:00 2001
From: Richard Nelson <wabz@pidgin.im>
Date: Mon, 6 Apr 2009 10:33:28 +0000
Subject: [PATCH] Set PURPLE_MESSAGE_RECV before receiving-im-msg. Also set it
 when receiving chat messages.

--HG--
extra : convert_revision : 0257ab6e534144a7fd317f997e1f0e19ff8bd82b
---
 libpurple/server.c |   16 +++++++++++-----
 1 file changed, 11 insertions(+), 5 deletions(-)

diff --git a/libpurple/server.c b/libpurple/server.c
index 6b04e1f..5da41a9 100644
--- a/libpurple/server.c
+++ b/libpurple/server.c
@@ -587,6 +587,11 @@ void serv_got_im(PurpleConnection *gc, const char *who, const char *msg,
 
 	account  = purple_connection_get_account(gc);
 
+	/*
+	 * XXX: Should we be setting this here, or relying on prpls to set it?
+	 */
+	flags |= PURPLE_MESSAGE_RECV;
+
 	if (PURPLE_PLUGIN_PROTOCOL_INFO(purple_connection_get_prpl(gc))->set_permit_deny == NULL) {
 		/* protocol does not support privacy, handle it ourselves */
 		if (!purple_privacy_check(account, who)) {
@@ -630,11 +635,6 @@ void serv_got_im(PurpleConnection *gc, const char *who, const char *msg,
 	if (conv == NULL)
 		conv = purple_find_conversation_with_account(PURPLE_CONV_TYPE_IM, name, gc->account);
 
-	/*
-	 * XXX: Should we be setting this here, or relying on prpls to set it?
-	 */
-	flags |= PURPLE_MESSAGE_RECV;
-
 	if (conv == NULL)
 		conv = purple_conversation_new(PURPLE_CONV_TYPE_IM, account, name);
 
@@ -925,6 +925,12 @@ void serv_got_chat_in(PurpleConnection *g, int id, const char *who,
 	g_return_if_fail(who != NULL);
 	g_return_if_fail(message != NULL);
 
+	/*
+	 * XXX: Should we be setting this here, or relying on prpls to set it?
+	 */
+	if (g_strcmp0(purple_account_get_username(g->account), who))
+		flags |= PURPLE_MESSAGE_RECV;
+
 	for (bcs = g->buddy_chats; bcs != NULL; bcs = bcs->next) {
 		conv = (PurpleConversation *)bcs->data;
 
-- 
1.7.10.4

